File: crm/app/internal/adapters/inbound/http/activity_handler.go
// Package http — Activity HTTP handlers.
// Auth: DEFERRED — LAN-trust only (per plan). No auth middleware here.
package http

import (
	"context"
	"encoding/json"
	"log"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/vantt/data-integration/crm/app/internal/domain"
)

// ActivityLister is the read interface needed by the activity handler.
type ActivityLister interface {
	ListTimeline(ctx context.Context, partyID string) ([]domain.Activity, error)
}

// ActivityLogger is the write interface needed by the activity handler.
type ActivityLogger interface {
	LogActivity(ctx context.Context, a *domain.Activity) error
}

// ActivityHandler wires activity endpoints.
type ActivityHandler struct {
	lister ActivityLister
	logger ActivityLogger
}

// NewActivityHandler constructs the handler.
func NewActivityHandler(lister ActivityLister, logger ActivityLogger) *ActivityHandler {
	return &ActivityHandler{lister: lister, logger: logger}
}

// RegisterRoutes mounts activity routes on the given chi router.
func (h *ActivityHandler) RegisterRoutes(r chi.Router) {
	r.Get("/api/parties/{id}/activities", h.listTimeline)
	r.Post("/api/parties/{id}/activities", h.logActivity)
}

// GET /api/parties/{id}/activities
func (h *ActivityHandler) listTimeline(w http.ResponseWriter, r *http.Request) {
	partyID := chi.URLParam(r, "id")
	acts, err := h.lister.ListTimeline(r.Context(), partyID)
	if err != nil {
		log.Printf("activity handler: list timeline %s: %v", partyID, err)
		writeError(w, http.StatusInternalServerError, "failed to list activities")
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"party_id": partyID, "activities": acts})
}

// POST /api/parties/{id}/activities
func (h *ActivityHandler) logActivity(w http.ResponseWriter, r *http.Request) {
	partyID := chi.URLParam(r, "id")

	var body struct {
		ActivityType     string  `json:"activity_type"`
		Direction        string  `json:"direction"`
		Channel          string  `json:"channel"`
		Subject          string  `json:"subject"`
		Body             string  `json:"body"`
		Outcome          string  `json:"outcome"`
		RelatedOrderCode string  `json:"related_order_code"`
		StaffUserID      *string `json:"staff_user_id"`
		OccurredAt       string  `json:"occurred_at"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	act := &domain.Activity{
		PartyID:          partyID,
		ActivityType:     body.ActivityType,
		Direction:        body.Direction,
		Channel:          body.Channel,
		Subject:          body.Subject,
		Body:             body.Body,
		Outcome:          body.Outcome,
		RelatedOrderCode: body.RelatedOrderCode,
		StaffUserID:      body.StaffUserID,
		OccurredAt:       body.OccurredAt,
	}
	if err := h.logger.LogActivity(r.Context(), act); err != nil {
		if domain.IsValidationError(err) {
			writeError(w, http.StatusBadRequest, err.Error())
		} else {
			log.Printf("activity handler: log activity %s: %v", partyID, err)
			writeError(w, http.StatusInternalServerError, "operation failed")
		}
		return
	}
	writeJSON(w, http.StatusCreated, act)
}

File: crm/app/adapters/inbound/http/activity_handler.py
"""HTTP adapter — Activity endpoints.

Auth: DEFERRED — LAN-trust only.
"""
from __future__ import annotations

import dataclasses
import logging

from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from pydantic import BaseModel

from application.activity_service import ActivityService

log = logging.getLogger(__name__)

router = APIRouter(prefix="/api")


# ── request schema ────────────────────────────────────────────────────────────

class LogActivityRequest(BaseModel):
    activity_type: str
    occurred_at: str = ""
    direction: str | None = None
    channel: str | None = None
    subject: str | None = None
    body: str | None = None
    outcome: str | None = None
    related_order_code: str | None = None
    staff_user_id: str | None = None


# ── dependency holder (set at app startup via wire_activity_router) ────────────

_svc: ActivityService | None = None


def wire_activity_router(svc: ActivityService) -> None:
    """Bind the application service; called during app startup."""
    global _svc
    _svc = svc


def _service() -> ActivityService:
    if _svc is None:
        raise RuntimeError("ActivityService not wired")
    return _svc


# ── endpoints ─────────────────────────────────────────────────────────────────

@router.post("/parties/{party_id}/activities", status_code=201)
def log_activity(party_id: str, req: LogActivityRequest):
    """POST /api/parties/{id}/activities — log a touchpoint for a party."""
    data = req.model_dump()
    data["party_id"] = party_id
    try:
        activity = _service().log_activity(data)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    except Exception as exc:
        log.error("activity handler: log_activity party=%s: %s", party_id, exc)
        raise HTTPException(status_code=500, detail="operation failed")
    return JSONResponse(status_code=201, content=dataclasses.asdict(activity))


@router.get("/parties/{party_id}/activities")
def list_timeline(party_id: str, limit: int = 50):
    """GET /api/parties/{id}/activities — timeline newest-first."""
    try:
        acts = _service().list_activities(party_id, limit=limit)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    except Exception as exc:
        log.error("activity handler: list_timeline party=%s: %s", party_id, exc)
        raise HTTPException(status_code=500, detail="failed to list activities")
    return {"party_id": party_id, "activities": [dataclasses.asdict(a) for a in acts]}
